lib/Containers/MerkleTree.cpp

```diff
@@ -0,0 +1,633 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2019 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Dan Larkin-York
+////////////////////////////////////////////////////////////////////////////////
+
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+#include <cstring>
+#include <functional>
+#include <limits>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <shared_mutex>
+#include <stdexcept>
+#include <string>
+#include <string_view>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include <chrono>
+#include <thread>
+
+#include "MerkleTree.h"
+
+#include "Basics/debugging.h"
+#include "Basics/hashes.h"
+
+#include "Logger/LogMacros.h"
+
+namespace arangodb {
+namespace containers {
+
+template <std::size_t const BranchingBits, std::size_t const LockStripes>
+bool MerkleTree<BranchingBits, LockStripes>::Node::operator==(Node const& other) {
+  return ((this->count == other.count) && (this->hash == other.hash));
+}
+
+template <std::size_t const BranchingBits, std::size_t const LockStripes>
+constexpr std::size_t MerkleTree<BranchingBits, LockStripes>::nodeCountAtDepth(std::size_t maxDepth) {
+  return static_cast<std::size_t>(1) << (BranchingBits * maxDepth);
+}
+class TestNodeCountAtDepth : public MerkleTree<3, 64> {
+  static_assert(nodeCountAtDepth(0) == 1);
```

Comment:
So there's a trade-off here. Either we can keep the implementation in the cpp file and keep the test here (simple static_assert for a constexpr function), or we can move the implementation to the header and move the tests out. Personally I'm for keeping code in the cpp file since the tests are so simple. Thoughts?